lib/sysmon/reporter.ex

```diff
@@ -0,0 +1,187 @@
+defmodule Instruments.Sysmon.Reporter do
+  @moduledoc """
+  Since only one process can subscribe to system monitor events, the Reporter
+  acts as a relay for system monitor events, allowing multiple subscribers to
+  receive system monitor events.
+
+  On startup, the Reporter will subscribe to the system monitor events
+  configured in `:sysmon_events` in the `:instruments` application environment.
+  If no events are configured, the Reporter will not subscribe to any events.
+  """
+  use GenServer
+
+  require Logger
+
+  @type sysmon_event ::
+          {:long_gc, pos_integer()}
+          | {:long_schedule, pos_integer()}
+          | {:large_heap, pos_integer()}
+          | :busy_port
+          | :busy_dist_port
+
+  @type t :: %__MODULE__{
+          subscribers: %{reference() => pid()},
+          events: [sysmon_event()]
+        }
+
+  defstruct subscribers: Map.new(), events: []
+
+  def start_link(opts \\ []) do
+    GenServer.start_link(__MODULE__, opts, name: __MODULE__)
+  end
+
+  @doc """
+  Subscribes the provided pid to configured system monitor events.
+  """
+  @spec subscribe(pid()) :: :ok
+  def subscribe(pid \\ self()) do
+    GenServer.call(__MODULE__, {:subscribe, pid})
+  end
+
+  @doc """
+  Unsubscribes the provided pid from system monitor events.
+  """
+  @spec unsubscribe(pid()) :: :ok
+  def unsubscribe(pid \\ self()) do
+    GenServer.call(__MODULE__, {:unsubscribe, pid})
+  end
+
+  @doc """
+  Sets the system monitor events to subscribe to. If no events are provided, the Reporter will not register itself as the system monitor process.
+  """
+  @spec set_events([sysmon_event()]) :: :ok
+  def set_events(events) do
+    GenServer.call(__MODULE__, {:set_events, events})
+  end
+
+  @doc """
+  Returns the system monitor events the Reporter is subscribed to.
+  """
+  @spec get_events() :: [sysmon_event()]
+  def get_events() do
+    GenServer.call(__MODULE__, :get_events)
+  end
+
+  @impl true
+  def init(_) do
+    sysmon_events = Application.get_env(:instruments, :sysmon_events, [])
+    enable_sysmon(sysmon_events)
+
+    {:ok,
+     %__MODULE__{
+       events: sysmon_events
+     }}
+  end
+
+  @impl true
+  def handle_call({:subscribe, pid}, _from, %__MODULE__{} = state) do
+    existing = Map.values(state.subscribers)
+
+    state =
+      if Enum.member?(existing, pid) do
+        state
+      else
+        ref = Process.monitor(pid)
+
+        %__MODULE__{
+          state
+          | subscribers: Map.put(state.subscribers, ref, pid)
+        }
+      end
+
+    {:reply, :ok, state}
+  end
+
+  def handle_call({:unsubscribe, pid}, _from, %__MODULE__{} = state) do
+    ref = Map.filter(state.subscribers, fn {_, p} -> p == pid end) |> Map.keys()
+
+    state =
+      case ref do
```

Comment:
If one pid calls `subscribe` more than once, `ref` will end up being a list with more than one element, and no unsubscription will be performed here. I would rewrite this function to something like

```ex
  state = 
    state.subscribers
    |> Enum.filter(state.subscribers, fn {_, p} -> p == pid end)
    |> Enum.reduce(state, fn {ref, _pid}, state ->
        Process.demonitor(ref)
        %__MODULE__{state | subscribers: Map.delete(state.subscribers, ref)
    end)

{:reply, :ok, state}
```

Reply:
It's not possible for `pid` to be in the map more than once, `subscribe/1` checks if the pid is already subscribed, if it is, it no-ops.

Reply:
Ah you're right. I still think it's a bit cleaner to handle it as a list still but that's just my 2c. Won't block on it